[Merged by Bors] - feat(Analysis/SpecialFunctions/StirlingRobbins): Robbins' sharp stepwise bound for stirlingSeq

FaffyWaffles opened the pull request:
This PR refactors `Stirling.lean` to use Robbin's bound `log (stirlingSeq n) - log (stirlingSeq (n+1)) ≤ 1 / (12 n (n + 1))`.

The original PR was by @FaffyWaffles but the PR has now been adopted by @tb65536.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

When merging, all the commits will be squashed into a single commit
listing all co-authors.

Co-authors in the squash commit are gathered from two sources:

First, all authors of commits to this PR branch are included. Thus,
one way to add co-authors is to include at least one commit authored by
each co-author among the commits in the pull request. If necessary, you
may create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

Second, co-authors can also be listed in lines at the very bottom of
the commit message (that is, directly before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines
at the bottom of the commit message (before the `---`, and also before
any "Co-authored-by" lines) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

FaffyWaffles committed:
feat(Stirling): add Robbins' sharp stepwise bound for stirlingSeq

Introduces Mathlib.Analysis.SpecialFunctions.StirlingRobbins, proving the sharp Robbins bound for successive differences in the logarithm of the Stirling sequence. This improves upon the previous bound and formalizes a result noted as missing in Mathlib documentation.

FaffyWaffles committed:
Move namespace and open scoped statements below module doc

Relocated the `open scoped BigOperators Nat` and `namespace Stirling` statements to appear after the module docstring

FaffyWaffles committed:
Merge branch 'leanprover-community:master' into feat-robbins-stirling-bound

tb65536 commented on Mathlib/Analysis/SpecialFunctions/StirlingRobbins.lean:
```diff
@@ -0,0 +1,153 @@
+        -- Evaluate geometric series: ∑ r^j = 1/(1-r)
+        rw [tsum_geometric_of_lt_one (by positivity) (ratio_sq_lt_one hk)]
+        ring
+    _ = (1 : ℝ) / (12 * k * (k + 1)) := geom_ratio_identity hk
```
I think this can be proved more directly. I don't have time to clean this up more right now, but this should already streamline things significantly:
```suggestion
theorem log_stirlingSeq_diff_le {k : ℕ} :
    Real.log (stirlingSeq k) - Real.log (stirlingSeq (k + 1)) ≤ (1 : ℝ) / (12 * k * (k + 1)) := by
  rcases k with (_ | k)
  · suffices 0 ≤ Real.log (Real.exp 1 / Real.sqrt 2) by simpa
    apply Real.log_nonneg
    grw [one_le_div (by positivity), ← Real.add_one_le_exp, Real.sqrt_le_left (by positivity)]
    norm_num
  set r := ((1 : ℝ) / (2 * (k + 1) + 1)) ^ 2
  suffices HasSum (fun j ↦ r ^ (j + 1) / 3) ((1 : ℝ) / (12 * (k + 1 : ℕ) * ((k + 1 : ℕ) + 1))) by
    refine hasSum_le (fun j ↦ ?_) (log_stirlingSeq_diff_hasSum k) this
    suffices (3 : ℝ) ≤ 2 * (j + 1) + 1 by simpa [field, r]
    norm_cast
    grind
  simp only [pow_succ]
  convert ((hasSum_geometric_of_lt_one (r := r) (by positivity) (by
    simp only [r]
    exact_mod_cast ratio_sq_lt_one (Nat.le_add_left 1 k))).mul_right r).div_const 3 using 1
  simp [field, r]
  ring_nf
  simp [field]
  ring_nf
```

FaffyWaffles committed:
Update Mathlib/Analysis/SpecialFunctions/StirlingRobbins.lean

Co-authored-by: Thomas Browning <tb65536@users.noreply.github.com>

FaffyWaffles committed:
Merge branch 'master' into feat-robbins-stirling-bound

FaffyWaffles committed:
feat(Analysis/SpecialFunctions/Stirling): add Robbins' sharp stepwise bound

Add `log_stirlingSeq_diff_le`, which proves Robbins' sharp bound of 1/(12k(k+1))
for successive differences in the Stirling sequence. This improves upon the existing
bound of 1/(4n²) in `log_stirlingSeq_sub_log_stirlingSeq_succ`.

This proof uses a geometric series comparison and achieves the optimal constant
for this type of bound. The file structure is reorganized into three parts, with
Part 3 containing the new result.

Documentation is updated throughout to reference the new bound and include proper
citations to Robbins' original paper.

Closes #31366

FaffyWaffles committed:
Merge branch 'master' into feat-robbins-stirling-bound

FaffyWaffles commented on Mathlib/Analysis/SpecialFunctions/Stirling.lean:
```diff
@@ -285,4 +312,36 @@ theorem le_log_factorial_stirling {n : ℕ} (hn : n ≠ 0) :
+This improves upon the bound of 1/(4n²) given in `log_stirlingSeq_sub_log_stirlingSeq_succ`.
+-/
+theorem log_stirlingSeq_diff_le (k : ℕ) :
+    Real.log (stirlingSeq k) - Real.log (stirlingSeq (k + 1)) ≤ (1 : ℝ) / (12 * k * (k + 1)) := by
```
I completely agree that having `log_stirlingSeq_diff_le` replace or 
subsume `log_stirlingSeq_sub_log_stirlingSeq_succ` would be cleaner, and definitely considered it. I decided to take a cautious approach about 
modifying any existing lemmas in my first PR since I wasn't sure about downstream dependencies.

j-loreaux reviewed (commented):
When you finish fiddling and add the bibliography reference, ping me for a review. This will likely be an easy merge.

tb65536 committed:
Merge branch 'master' into feat-robbins-stirling-bound

tb65536 committed:
Merge branch 'master' into feat-robbins-stirling-bound

tb65536 committed:
move

tb65536 committed:
tweak

tb65536 committed:
tweak

tb65536 committed:
tweak

j-loreaux reviewed (commented):
Thanks!

bors merge
